Fatras/Plugins/Geant4/src/Decay.cpp

```diff
@@ -0,0 +1,99 @@
+
+#include "ActsFatras/Plugins/Geant4/Decay.hpp"
+
+#include "Acts/Definitions/Common.hpp"
+#include "ActsFatras/EventData/ProcessType.hpp"
+#include "ActsFatras/Plugins/Geant4/G4DetectorConstruction.hpp"
+
+#include "G4DecayProducts.hh"
+#include "G4DecayTable.hh"
+#include "G4RunManager.hh"
+#include "QGSP_BERT.hh"
+
+ActsFatras::Decay::Decay() : m_g4RunManager(initG4RunManager()) {}
+
+std::vector<ActsFatras::Particle> ActsFatras::Decay::decayParticle(
+    const ActsFatras::Particle& parent) const {
+  std::vector<Particle> children;
+
+  // Find the particle type that will decay
+  int pdgCode = parent.pdg();
+  G4ParticleDefinition* pDef = m_pdgToG4Conv.getParticleDefinition(pdgCode);
+  if (!pDef) {
+    return children;
+  }
+  // Get the particles decay table
+  G4DecayTable* dt = pDef->GetDecayTable();
+  if (!dt) {
+    return children;
+  }
+  // Select a decay channel
+  G4VDecayChannel* channel = dt->SelectADecayChannel();
+  if (!channel) {
+    return children;
+  }
+  // Get the decay products from the selected channel
+  G4DecayProducts* products = channel->DecayIt();
+  if (!products) {
+    return children;
+  }
+
+  // Boost the decay products using the parents four-momentum
+  const Particle::Vector4 mom4 = parent.fourMomentum();
+  products->Boost(mom4[Acts::eMom0] / mom4[Acts::eEnergy],
+                  mom4[Acts::eMom1] / mom4[Acts::eEnergy],
+                  mom4[Acts::eMom2] / mom4[Acts::eEnergy]);
+
+  G4int nProducts = products->entries();
+  for (G4int i = 0; i < nProducts; i++) {
+    G4DynamicParticle* prod = products->PopProducts();
+    if (!prod) {
+      continue;
+    }
+
+    // Convert the decay product from Geant4 to Acts
+    const G4ThreeVector& mom = prod->GetMomentum();
+    constexpr Scalar convertEnergy = Acts::UnitConstants::GeV / CLHEP::GeV;
+    Acts::Vector3 amgMom(mom.x(), mom.y(), mom.z());
+    amgMom *= convertEnergy;
+    const int32_t pdg = prod->GetPDGcode();
+
+    Particle childParticle(Barcode(), static_cast<Acts::PdgParticle>(pdg));
+    childParticle.setPosition4(parent.fourPosition())
+        .setAbsoluteMomentum(amgMom.norm())
+        .setDirection(amgMom)
+        .setProcess(ProcessType::eDecay);
+
+    // Store the particle
+    children.push_back(std::move(childParticle));
+  }
+  return children;
+}
+
+G4RunManager* ActsFatras::Decay::initG4RunManager() const {
+  // Test if there's already a G4RunManager
+  if (G4RunManager::GetRunManager() == nullptr) {
+    G4RunManager* runManager = new G4RunManager;
+
+    // Initialise physics
+    G4VUserPhysicsList* thePL = new QGSP_BERT;
+    runManager->SetUserInitialization(thePL);
+
+    // Build a dummy detector
+    runManager->SetUserInitialization(new G4DetectorConstruction());
+
+    // Initialise the G4RunManager itself
+    runManager->Initialize();
+    return runManager;
+  } else {
+    // Return the existing G4RunManager
+    return G4RunManager::GetRunManager();
+  }
+}
```

Comment:
This has nothing to do with the decay implementation. As previously suggested, I think this would be best placed as a standalone function `makeDummyRunManager` in a separate header file. That would also allow the dummy detector construction to become a purely private implementation detail that does not need to have a public header.

To answer your question from a related previous thread: the decay module should then have a single constructor `Decay(G4RunManager*)` that always takes an user-provided pointer. In most cases, this can be the return value from `makeDummyRunManager` but it could but something else for more complicated setups. The constructor must throw in case of a null ptr, but then all member calls can assume that the run manager pointer is valid.

Reply:
I see your point in the first paragraph.
About the second: The decay object must be default constructable, so you will have to call this method in any case. If you want an additional constructor in order to pass this decay class with you run manager to the `Simulator`/ `Interactor` then you will end up with the same `G4RunManager` you had in the default constructor since this is `static` object. So while the current strucure remains it is pointless. Furthermore, there is no such thing as a "more complicated setup" in the context of the particle decay. This manager class is just required for G4 to work (i.e. you can't have physics without detector setup etc. / internal G4 logic).  In our case we need the tables.

Reply:
I forgot about the default-constructible requirement. Yes, then this is the only way to implement it. The updated implementation looks good.

The more *complicated setup* I was thinking about was an additional Geant4-based hadronic interaction module. In that case, both should share the run manager. But since Geant4 already has the concept of a global run manager, this is probably not actually a problem.

Reply:
For now it isn't a problem but it might come back later (once this module is available) as I don't know by heart whether this implies a certain run manager initialisation.

Reply:
I think that is good enough for now.